CodeDriven Unit localization server with AttributePersistenceMigration

ratgr opened the pull request:
#### Summary

Reimplementation of UnitLocalizationCluster as a CodeDriven cluster
Uses the AttributePersistenceMigration to implement UnitLocalization keeping old data in the mix

#### Related issues
Pr with only the persistence migration changes: 
Implements: #41179

#### Testing
Changed UnitLocalization server attributes on the old app
Updated and run new app, old values are kept

 
#### Readability checklist

The checklist below will help the reviewer finish PR review in time and keep the
code readable:

-   [X] PR title is
        [descriptive](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html#title-formatting)
-   [X] Apply the
        [_“When in Rome…”_](https://project-chip.github.io/connectedhomeip-doc/style/CODING_STYLE_GUIDE.html)
        rule (coding style)
-   [x] PR size is short
-   [X] Try to avoid "squashing" and "force-update" in commit history
     - Accidental update of submodules needed a force-update to remove   
-   [C] CI time didn't increase

See: [Pull Request Guidelines](https://project-chip.github.io/connectedhomeip-doc/contributing/pull_request_guidelines.html)

ratgr committed:
Add migration

ratgr committed:
ImproveInterface for the expected usage

restyled-commits committed:
Restyled by clang-format

ratgr committed:
Fix CI issues

ratgr committed:
quick fix

ratgr committed:
more fixes

restyled-commits committed:
Restyled by clang-format

ratgr committed:
missed fix

ratgr committed:
Fixes on bigendianness

restyled-commits committed:
Restyled by clang-format

ratgr committed:
Auto instead of StorageKeyName

restyled-commits committed:
Restyled by clang-format

ratgr committed:
fixing broken CI

restyled-commits committed:
Restyled by clang-format

ratgr committed:
Adding dependencies

restyled-commits committed:
Restyled by clang-format

ratgr committed:
StorageKeyName not found

restyled-commits committed:
Restyled by clang-format

ratgr committed:
adding dependency

ratgr committed:
typo fix

ratgr committed:
more changes

ratgr committed:
Added the delete method to Attribute Persistance

restyled-commits committed:
Restyled by clang-format

ratgr committed:
fix error on virtual fucntion

ratgr committed:
Adding more deps

ratgr committed:
Merge branch 'master' into compatible-attrpersistence

ratgr committed:
Separating the Migration Code

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

ratgr committed:
Merge branch 'master' into compatible-attrpersistence

ratgr committed:
Merge branch 'master' into compatible-attrpersistence

ratgr committed:
Merge branch 'master' into compatible-attrpersistence

ratgr committed:
Merge branch 'master' into compatible-attrpersistence

ratgr committed:
Adding Cpp files into CMake system

ratgr committed:
Merge branch 'master' into compatible-attrpersistence

ratgr committed:
Merge branch 'master' into compatible-attrpersistence

ratgr committed:
clangtidy fix

ratgr committed:
Removing basic information

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Code driven Unit Localization Server

ratgr committed:
missing piece

ratgr committed:
Merge remote-tracking branch 'ratgr/unit-localization-server' into unit-localization-server

ratgr committed:
Updates to persistence

ratgr committed:
Unit Localization fixing the CodegenIntegration

ratgr committed:
delaying the rename of unit-localization-server

ratgr committed:
Removing the renaming

ratgr committed:
missing change for the renaming

ratgr committed:
remove persistence migration from linter

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Changing attributte generation

ratgr committed:
using listbuilder for the attributes

ratgr committed:
adding attributes to zcl

restyled-commits committed:
Restyled by whitespace

ratgr committed:
more changes for migration

ratgr committed:
reorder zcl

ratgr committed:
surround migration code with ndef

ratgr committed:
Fixing ID

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
unit tests unit loclaization cluster

ratgr committed:
Merge branch 'master' into unit-localization-server

andy31415 commented on src/app/clusters/unit-localization-server/unit-localization-server.cpp:
```diff
@@ -16,49 +14,31 @@
-#include <app/reporting/reporting.h>
-#include <lib/support/CodeUtils.h>
-#include <lib/support/logging/CHIPLogging.h>
+#ifndef CHIP_SKIP_PERSISTENCE_MIGRATION
```
we likely want to add it in an existing buildconfig header at least, so that a gn config propagates to a header and then we use that value. should then use `#if` and never ifdef/defined to ensure that we explicitly set a value to something.

andy31415 replied:
I don't think we have a very sane design currently. What I would want is to have libCHIP a stand-alone library that does not vary by app, which means such a "migrate" flag should be part of an app config (which should not reside in libCHIP folders like src/app but rather somewhere in the application specific configuration)

I think that we need migration of "old" code and that is generally ember framework and ember framework & CodegenIntegration is always app specific - we should place things there.

I propose:
  - migration logic that is generic can be part of libCHIP (LTO will optimize it out if not needed)
  - actual cluster migration should be in CodegenIntegration.cpp (or codegenintegration should be the only thing calling some migration function). This is because codegenintegration can make use of app specific configs

Header wise I see nothing obvious, however we see to have:
  - all examples have a CHIPProjectAppConfig
  - those are included via `chip_device_project_config_include`/`CHIP_DEVICE_PROJECT_CONFIG_INCLUDE` in `src/include/platform/CHIPDeviceConfig.h`
  - https://github.com/project-chip/connectedhomeip/blob/master/src/platform/BUILD.gn#L116 defines that header 
  
So we can add it as a gn argument (that apps can use) in CHIPDeviceBuildConfig.h and apps will allso be able to override it in their project config. I find this not great (platforms feels like a questionable location), however it should allow for app-specific overriding.

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr commented on src/app/persistence/AttributePersistenceMigration.h:
```diff
@@ -0,0 +1,82 @@
+    DefaultAttributePersistenceProvider normProvider;
+    normProvider.Init(&storageDelegate);
+
+    unsigned char attributeBuffer[attributeBufferSize] = {};
```
Fixed

ratgr commented on src/app/clusters/unit-localization-server/unit-localization-server.cpp:
```diff
@@ -74,80 +54,78 @@ CHIP_ERROR UnitLocalizationServer::SetSupportedTemperatureUnits(DataModel::List<
+        return encoder.Encode(mTemperatureUnit);
+
+    case UnitLocalization::Attributes::SupportedTemperatureUnits::Id:
+        return encoder.Encode(mSupportedTemperatureUnits);
```
Perf will do

ratgr commented on src/app/clusters/unit-localization-server/unit-localization-server.cpp:
```diff
@@ -74,80 +54,78 @@ CHIP_ERROR UnitLocalizationServer::SetSupportedTemperatureUnits(DataModel::List<
-    }
-    return CHIP_NO_ERROR;
+
+    return listBuilder.Append(Span<DataModel::AttributeEntry>(), Span(optionalAttributes), optionalAttributeSet);
```
Got you! Will fix

ratgr committed:
generation

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
build gn all clusters app add unitlocalization

ratgr committed:
Merge remote-tracking branch 'origin' into unit-localization-server

ratgr committed:
merge issues

ratgr committed:
Separating migration to CodeGenIntegration

ratgr committed:
nullptr

ratgr committed:
reorder

restyled-commits committed:
Restyled by clang-format

ratgr committed:
remove ifdefs

restyled-commits committed:
Restyled by clang-format

ratgr committed:
separating migration call

ratgr committed:
fix

ratgr committed:
make migration more const friendly

ratgr committed:
add more const

ratgr committed:
New migration

ratgr committed:
Merge remote-tracking branch 'origin' into unit-localization-server

ratgr committed:
Quick merge fix

ratgr committed:
make UnitLocalizationClusterWithMigration members public

ratgr committed:
making logs consistent to optimze away some strings

ratgr committed:
fix for more platforms

ratgr committed:
Change log values

ratgr committed:
reducing logs

ratgr committed:
more log changes

ratgr committed:
Changing to PRI values

ratgr committed:
using ChipLogValueMEI

ratgr committed:
Unifiying log format

ratgr committed:
Moving Migration log format to a define
I'm moving the log format into a define, using the constexpr didn't work on all platforms so now we need it as a define

ratgr committed:
ESP32 partition changes
Making ota_0 bigger so that the all-cluster app can fit, and removing ota_1 as recommended by express_if team

ratgr committed:
Fix missing CHIP_ERROR_FORMAT in K_MIGRATION_LOG_FORMAT

ratgr committed:
Merge remote-tracking branch 'origin' into unit-localization-server

ratgr committed:
Fix unit test

ratgr committed:
Merge branch 'master' into unit-localization-server

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ratgr committed:
Merge with local

ratgr committed:
fix merge issue

ratgr committed:
Fixes ignoring return

ratgr committed:
Revert the changed submodules

This reverts commit 3b5252b3479ba1ef718c46ec8e6e3a08e8617994.

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
fix returns again

ratgr committed:
Return error if migration fails

ratgr committed:
regen zap

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Partial Revert "regen zap",

 Changed MTRBaseClusters and CI doesn't like the change

This reverts commit d4e21a89885b47e106fa0ae67f6325dc182e56ea.

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Changes to zap file

ratgr committed:
Removing condition from LUnit_3_1

restyled-commits committed:
Restyled by ruff

restyled-commits committed:
Restyled by isort

ratgr committed:
Partial Revert "Changes to zap file", reverts changes not for unit localization

This reverts commit d487e9af6a7b1a160e6fe525728b67b1c45d3288.

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Revert "Changes to zap file"

This reverts commit d487e9af6a7b1a160e6fe525728b67b1c45d3288.

ratgr committed:
Fixing feature map on all-clusters-app

restyled-commits committed:
Restyled by prettier-json

ratgr committed:
Generation fixes

ratgr committed:
Fixing zap gen

ratgr committed:
revert change in python test

ratgr committed:
Revertomg condition from LUnit_3_1

ratgr committed:
Revert "Revertomg condition from LUnit_3_1"

This reverts commit 800f174d07b7db55ef941b8945cf6198a112df72.

ratgr committed:
adding has_attribute to python includes

restyled-commits committed:
Restyled by isort

ratgr commented on examples/all-clusters-app/all-clusters-common/all-clusters-app.matter:
```diff
@@ -7761,8 +7761,8 @@ endpoint 0 {
     callback attribute generatedCommandList;
     callback attribute acceptedCommandList;
     callback attribute attributeList;
-    ram      attribute featureMap default = 0x1;
```
Fixed

ratgr commented on src/python_testing/TC_LUNIT_3_1.py:
```diff
@@ -82,7 +81,7 @@ def pics_TC_LUNIT_3_1(self) -> list[str]:
         ]
 
-    @run_if_endpoint_matches(has_cluster(Clusters.UnitLocalization) and has_attribute(Clusters.UnitLocalization.Attributes.TemperatureUnit))
+    @run_if_endpoint_matches(has_cluster(Clusters.UnitLocalization))
```
Fixed

ratgr committed:
Change Test for Unit Localization w/Migration

ratgr committed:
doing startup on MIgration

ratgr committed:
reenable migration

ratgr committed:
Revert Submodule changes

This reverts commit 7aabe64ce094138db6bc3a9ad2944c0a61f9b0e1.

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Fixing test execution

ratgr committed:
continue

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
localization cluster unit tests fix

ratgr committed:
Merge remote-tracking branch 'origin/master' into unit-localization-server

ratgr committed:
Revert "localization cluster unit tests fix"

This reverts commit 9e70e3e1d221ede41c8d2a13b404f05f4505cb09.

ratgr committed:
Revert "Merge remote-tracking branch 'origin/master' into unit-localization-server"

This reverts commit 4826e2055843c3b1b78125ba1fd53b991c818a98, reversing
changes made to 9e70e3e1d221ede41c8d2a13b404f05f4505cb09.

ratgr committed:
Fixing tests

ratgr committed:
Reapply "Merge remote-tracking branch 'origin/master' into unit-localization-server"

This reverts commit 1a3f3e38519c9a2ff5e2815ea240ba6563288890.

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Reapply Submodules from  "Merge remote-tracking branch 'origin/master' into unit-localization-server"

This reverts commit 1a3f3e38519c9a2ff5e2815ea240ba6563288890.

ratgr committed:
Revert simw-top-mini  "Merge remote-tracking branch 'origin/master' into unit-localization-server""

This reverts commit fefa55f7f6851c476a842a3369730c5e866c5959.

ratgr committed:
simw-top-mini revert try 2

ratgr committed:
remove using namespace chip::Test

ratgr committed:
remove unnecesary include

ratgr committed:
change persistence migration to include a function to get the value

ratgr committed:
remove unnecesary header in codeIntegration

ratgr committed:
fixes in migration data

ratgr committed:
make span a const pair

ratgr committed:
add CHIP_ERROR return

ratgr committed:
use underlying type

ratgr committed:
Adding const

ratgr committed:
build fixes

andy31415 commented on src/app/clusters/unit-localization-server/CodeGenIntegration.cpp:
```diff
@@ -0,0 +1,94 @@
+
+void MatterUnitLocalizationClusterInitCallback(EndpointId endpointId)
+{
+    IntegrationDelegate integrationDelegate;
```
should we check that endpointId == kRootEndpointId? other singletons do this: https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/basic-information/CodegenIntegration.cpp#L74

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
Merge logs

ratgr committed:
Revert "Merge logs"

This reverts commit af316c6718662dd3078dbb00831f5f2a67328514.

ratgr committed:
optimizing migration

ratgr committed:
Merge branch 'master' into unit-localization-server

ratgr committed:
make migration not generate moves

ratgr committed:
Merge branch 'master' into unit-localization-server
